t/SocketWithLogging.hpp

```diff
@@ -3,76 +3,95 @@
 #include <boost/asio/buffers_iterator.hpp>
 #include <boost/system/error_code.hpp>
 #include <iostream>
+#include <string>
 #include <utility>
-#include <mutex>
 
 namespace bredis {
 namespace test {
 
 template <typename NextLayer> class SocketWithLogging {
     NextLayer stream_;
-    std::mutex cout_mutex_;
+
+    template <typename BufferSequence, typename Handler>
+    struct HandlerWithLogging {
+        Handler next_handler_;
+        const BufferSequence buffers_;
+        const char *prefix_;
+
+        HandlerWithLogging(const char *prefix, Handler hander,
+                           const BufferSequence &buffers)
+            : prefix_(prefix), next_handler_(hander), buffers_(buffers) {}
+
+        void operator()(const boost::system::error_code &error_code,
+                        std::size_t bytes_transferred) {
+            dump(prefix_, buffers_, bytes_transferred);
+            next_handler_(error_code, bytes_transferred);
+        }
+    };
 
   public:
     template <typename... Args>
     SocketWithLogging(Args &&... args) : stream_(std::forward<Args>(args)...) {}
 
     template <typename BufferSequence>
-    void dump(const char *prefix, const BufferSequence &buffers) {
-        using Iterator = boost::asio::buffers_iterator<BufferSequence, char>;
-
-        Iterator it = Iterator::begin(buffers);
-        Iterator end = Iterator::end(buffers);
-        auto size = std::distance(it, end);
+    static void dump(const char *prefix, const BufferSequence &buffers,
+                     std::size_t size) {
+        using boost::asio::buffer_cast;
+        using boost::asio::buffer_size;
 
-        std::lock_guard<std::mutex> session_lock(cout_mutex_);
-        std::cout << "{" << prefix << " " << size << " bytes}[";
-        for (; it != end; it++) {
-            std::cout << *it;
+        std::string content;
+        content.reserve(size);
+        for (auto const &buffer : buffers) {
+            content.append(buffer_cast<char const *>(buffer), size);
```

Comment:
This is inefficient, instead of appending to a string and then dumping the string, why not just loop and stream? Like this:
```
for (auto const &buffer : buffers)
  std::cout << buffer_cast<char const*>(buffer);
```
You have to add code that counts how many bytes are going out, and account for only writing `size` bytes. But this would be more efficient.